airbyte-cdk/python/airbyte_cdk/destinations/vector_db_based/writer.py

```diff
@@ -29,20 +29,25 @@ def __init__(self, processing_config: ProcessingConfigModel, indexer: Indexer, e
         self.processing_config = processing_config
         self.indexer = indexer
         self.embedder = embedder
-        self.batcher = Batcher(batch_size, lambda batch: self._process_batch(batch))
-
-    def _process_batch(self, batch: List[AirbyteRecordMessage]) -> None:
-        documents: List[Chunk] = []
-        ids_to_delete = []
-        for record in batch:
-            record_documents, record_id_to_delete = self.processor.process(record)
-            documents.extend(record_documents)
-            if record_id_to_delete is not None:
-                ids_to_delete.append(record_id_to_delete)
-        embeddings = self.embedder.embed_chunks(documents)
-        for i, document in enumerate(documents):
-            document.embedding = embeddings[i]
-        self.indexer.index(documents, ids_to_delete)
+        self.batch_size = batch_size
+        self._init_batch()
+
+    def _init_batch(self) -> None:
+        self.documents: Dict[Tuple[str, str], List[Chunk]] = defaultdict(list)
+        self.ids_to_delete: Dict[Tuple[str, str], List[str]] = defaultdict(list)
+        self.number_of_documents = 0
+
+    def _process_batch(self) -> None:
+        for (namespace, stream), ids in self.ids_to_delete.items():
+            self.indexer.delete(ids, namespace, stream)
+
+        for (namespace, stream), documents in self.documents.items():
+            embeddings = self.embedder.embed_chunks(documents)
+            for i, document in enumerate(documents):
+                document.embedding = embeddings[i]
+            self.indexer.index(documents, namespace, stream)
```

Comment:
@flash1293 - To protect against queries hitting the index after the delete operation and before the insert - are you finding that at least some vectorstores support transactions?

I'm considering if `_process_batch()` should call `start_transaction()` and `commit_transaction()` on the indexer at the top and bottom of this code block. Even as a no-op, the subclasses could then implement internally (or not) without requiring us to modify the central CDK code.

It's okay to defer solving this, I think, but I at least wanted to start the conversation...

Reply:
They usually don't, it sounds like this is very similar to the idea in https://github.com/airbytehq/airbyte/issues/30281

If we would use transactions on this level we would guarantee that the documents belonging to each record would not disappear for a short amount of time between the delete and the load, but it would still have the issue of half of the dataset being outdated while half of the dataset being up to date while the sync is running.